### PR TITLE
Add the IDDES Model

### PR DESCRIPTION
**Pull request by psakievich**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [x] Feature enhancement

## Description of the pull-request

This PR is to merge the IDDES model into master.  I will open as a draft and move it to a full PR once the conflicts are removed.

TODO:
- [x] Add support for legacy blending and Strelets blending
- [x] Add back in legacy open BC behavior
- [x] Add unit tests
- [ ] Add regression test

## Checklist

*All pull requests*
- [x] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [ ] Linux
    - [x] MacOS
  - Compilers 
    - [ ] GCC
    - [x] LLVM/Clang
    - [ ] Intel compilers
    - [ ] NVIDIA CUDA
- [ ] Compiles without warnings
- [ ] Passes all unit tests
- [ ] Passes all regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [ ] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [ ] New regression tests exercising the new code

### CODE REVIEW

File: src/LowMachEquationSystem.C
Comment (psakievich) [inline src/LowMachEquationSystem.C:1184]: @gantech do you have any heart burn with me removing these fields?  It looks like they are only used for the Strelets upwinding computation and are a bunch of intermediate values.

File: include/edge_kernels/MomentumEdgePecletAlg.h
Comment (sayerhs) [inline include/edge_kernels/MomentumEdgePecletAlg.h:26]: Comment indicates you wanted to use SIMD here?

File: include/ngp_algorithms/EffSSTIDDESABLDiffFluxCoeffAlg.h
Comment (sayerhs) [inline include/ngp_algorithms/EffSSTIDDESABLDiffFluxCoeffAlg.h:2]: Fix license

File: include/ngp_algorithms/TurbViscSSTIDDESABLAlg.h
Comment (sayerhs) [inline include/ngp_algorithms/TurbViscSSTIDDESABLAlg.h:2]: Fix license

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (sayerhs) [inline include/node_kernels/SDRSSTIDDESABLNodeKernel.h:2]: Fix license

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (sayerhs) [inline include/node_kernels/SDRSSTIDDESABLNodeKernel.h:28]: This will generate warnings on AppleClang because `const int nDim_` isn't initialized... so this constructor will be deleted.

File: include/node_kernels/TKESSTIDDESNodeKernel.h
Comment (sayerhs) [inline include/node_kernels/TKESSTIDDESNodeKernel.h:2]: Fix license

File: src/LowMachEquationSystem.C
Comment (sayerhs) [inline src/LowMachEquationSystem.C:1042]: Do we want this in this pull-request? We might want to rethink this than hard code this in LowMachEQS. My preference would be to take this out for IDDES PR

File: src/LowMachEquationSystem.C
Comment (sayerhs) [inline src/LowMachEquationSystem.C:1181]: Same comment as above

File: src/ShearStressTransportEquationSystem.C
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C:321]: remove this as this was recently addressed in #174 and af5102224fa71160

```suggestion
```

File: src/TurbKineticEnergyEquationSystem.C
Comment (sayerhs) [inline src/TurbKineticEnergyEquationSystem.C:171]: Looks like it is time to refactor these if conditions into a function :)

File: src/TurbKineticEnergyEquationSystem.C
Comment (sayerhs) [inline src/TurbKineticEnergyEquationSystem.C:351]: Looks like this will never happen as the constructor already throws

File: src/TurbKineticEnergyEquationSystem.C
Comment (sayerhs) [inline src/TurbKineticEnergyEquationSystem.C:497]: same comment as above

File: src/edge_kernels/MomentumEdgeSolverAlg.C
Comment (sayerhs) [inline src/edge_kernels/MomentumEdgeSolverAlg.C:75]:
```suggestion

```

File: src/edge_kernels/StreletsUpwindEdgeAlg.C
Comment (sayerhs) [inline src/edge_kernels/StreletsUpwindEdgeAlg.C:42]: Solution options has been initialized before algorithms are created.

File: src/ngp_algorithms/EffSSTIDDESABLDiffFluxCoeffAlg.C
Comment (sayerhs) [inline src/ngp_algorithms/EffSSTIDDESABLDiffFluxCoeffAlg.C:2]: Fix license

File: src/ngp_algorithms/TurbViscSSTIDDESABLAlg.C
Comment (sayerhs) [inline src/ngp_algorithms/TurbViscSSTIDDESABLAlg.C:2]: Fix license

File: src/node_kernels/SDRSSTIDDESABLNodeKernel.C
Comment (sayerhs) [inline src/node_kernels/SDRSSTIDDESABLNodeKernel.C:2]: Fix license

File: src/node_kernels/TKESSTIDDESABLNodeKernel.C
Comment (sayerhs) [inline src/node_kernels/TKESSTIDDESABLNodeKernel.C:2]: Fix license

File: src/node_kernels/TKESSTIDDESNodeKernel.C
Comment (sayerhs) [inline src/node_kernels/TKESSTIDDESNodeKernel.C:2]: Fix license

**Review by sayerhs — COMMENTED** [review]
This PR should not change the submodule snapshot. Looks like it is removing the recent meshes that are required for newer AMS regression tests.

File: include/edge_kernels/MomentumEdgePecletAlg.h
Comment (psakievich) [reply include/edge_kernels/MomentumEdgePecletAlg.h:26]: yeah sorry this is a stale comment.  I'll delete it.

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (psakievich) [reply include/node_kernels/SDRSSTIDDESABLNodeKernel.h:28]: Ah good catch!

File: src/ShearStressTransportEquationSystem.C
Comment (psakievich) [reply src/ShearStressTransportEquationSystem.C:321]: 👍

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (psakievich) [reply include/node_kernels/SDRSSTIDDESABLNodeKernel.h:28]: @sayerhs do you have a preferred way of dealing with this?  I'd be fine with deleting the default constructor as long as we don't need it for the device. Seems like there are several of these kicking around based off my compiler warnings on the laptop, and I'd like to clean them up (in a separate PR) 😄

File: src/LowMachEquationSystem.C
Comment (psakievich) [reply src/LowMachEquationSystem.C:1042]: Yeah you are right. This is pretty ugly.  I'll pull it into a separate function.

File: src/TurbKineticEnergyEquationSystem.C
Comment (psakievich) [reply src/TurbKineticEnergyEquationSystem.C:351]: Would your rather remove the default switch statement altogether?  I kind of like leaving it in, but I'd change the error message to not list the models since it should serve as a developer warning.  I'm fine removing it altogether if you'd prefer though.

File: Debug
Comment (sayerhs) [inline Debug:1]: Remove development files from PR

File: Release
Comment (sayerhs) [inline Release:1]: Remove development files from PR

File: ReleaseWhOpenFAST
Comment (sayerhs) [inline ReleaseWhOpenFAST:1]: Remove development files from PR

**Review by sayerhs — CHANGES_REQUESTED** [review]
@psakievich I don't know of a better way to block accidental merges than choosing _request changes_. Please don't merge this until you've removed all your build files from the PR

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (sayerhs) [reply include/node_kernels/SDRSSTIDDESABLNodeKernel.h:28]: @psakievich removing the default constructor is the right way to fix this (as we can't provide a value for `nDim_` without consulting Realm). We don't need the default constructor for the device as we are not creating a `Kokkos::View` of the `NodeKernel` but through a holder struct.

File: src/TurbKineticEnergyEquationSystem.C
Comment (sayerhs) [reply src/TurbKineticEnergyEquationSystem.C]: I don't think we can remove the `default` because it would generate compile time warnings. How about adding just a comment so that developers know what is going on? 

```C++
default:
  // Should never reach here as we check and throw in constructor
  break;
```

File: include/node_kernels/SDRSSTIDDESABLNodeKernel.h
Comment (psakievich) [reply include/node_kernels/SDRSSTIDDESABLNodeKernel.h:28]: :+1: thanks for clarifying that.  I wasn't sure how these kernels were being sent to device.

**Comment by psakievich** [comment]
@sayerhs sorry about the build files.  I forgot I'd put some symlinks in place.  I've removed them.

File: src/TurbKineticEnergyEquationSystem.C
Comment (psakievich) [reply src/TurbKineticEnergyEquationSystem.C]: Sure that works for me.

**Comment by sayerhs** [comment]
@psakievich The Strelets upwind unit test is failing on CI (https://github.com/Exawind/nalu-wind/pull/725/checks) is this expected?

**Comment by psakievich** [comment]
@sayerhs yeah I haven't finished it yet. It's still WIP. Once that is done this should be ready to merge

**Review by rcknaus — COMMENTED** [review]
Looks good overall.  For the different SST flavors, we really just need to compute a different length scale (or destruction term, alternatively), so consider lifting the length scale computation out as a function or maybe writing an algorithm that writes the length-scale to a field. Not really critical though.

**Review by sayerhs — APPROVED** [review]

File: src/edge_kernels/MomentumOpenEdgeKernel.C
Comment (sayerhs) [reply src/edge_kernels/MomentumOpenEdgeKernel.C:223]: IDDES_ABL is a blending between ABL length scales and the LES part of DES model. So there are two levels of blending going on, and they are two separate models.

File: src/MiscIDDESABL.C
Comment (psakievich) [reply src/MiscIDDESABL.C:43]: That's good to know @sayerhs. I wasn't aware of that.  I don't really know all the details of this implementation. I'm just trying to get it mergeable at this point.  Should I just delete this function then? @gantech want to weigh in?

File: src/LowMachEquationSystem.C
Comment (gantech) [reply src/LowMachEquationSystem.C]: Thanks @psakievich Please go ahead and remove them. As long as we have a final working version of IDDES model... I'm more than happy and really glad for the help.

File: src/MiscIDDESABL.C
Comment (gantech) [reply src/MiscIDDESABL.C:43]: Please go ahead and delete this if necessary. I can add this capability later.

File: src/edge_kernels/StreletsUpwindEdgeAlg.C
Comment (psakievich) [reply src/edge_kernels/StreletsUpwindEdgeAlg.C:47]: @rcknaus that was the intent here.  Are you thinking we check and throw in solution options or something? I was also thinking we could add strelets upwinding as an option within solution options at some point so it could be used with sst_des as well

File: src/node_kernels/TKESSTIDDESABLNodeKernel.C
Comment (psakievich) [reply src/node_kernels/TKESSTIDDESABLNodeKernel.C:19]: Yeah I really like this idea.  I don't think I will have time to do this on this PR though unless someone else wants to jump in.  I need to wrap this one up pretty soon.

File: src/MiscIDDESABL.C
Comment (psakievich) [reply src/MiscIDDESABL.C:43]: Okay, yeah I'll remove this for now.  I think we should make this interface cleaner. Perhaps add it as a a restart flag and/or add some parameters to control the blending.

File: src/edge_kernels/StreletsUpwindEdgeAlg.C
Comment (rcknaus) [reply src/edge_kernels/StreletsUpwindEdgeAlg.C:47]: Yeah, more or less.  Like having strelets on automatically switches the blending coefficients to the values they have to be, and we check that the user isn't setting them to something else in solution options.

File: src/node_kernels/TKESSTIDDESABLNodeKernel.C
Comment (rcknaus) [reply src/node_kernels/TKESSTIDDESABLNodeKernel.C:19]: Yeah, that's fine.  Just wanted to make the suggestion.

File: src/MiscIDDESABL.C
Comment (psakievich) [reply src/MiscIDDESABL.C:24]: :+1:

**Review by gantech — APPROVED** [review]

File: src/MiscIDDESABL.C
Comment (psakievich) [reply src/MiscIDDESABL.C:102]: Removing SST_IDDES_ABL for now.

File: src/node_kernels/SDRSSTIDDESABLNodeKernel.C
Comment (psakievich) [reply src/node_kernels/SDRSSTIDDESABLNodeKernel.C]: Removed SST_IDDES_ABL from this PR

File: src/edge_kernels/StreletsUpwindEdgeAlg.C
Comment (psakievich) [reply src/edge_kernels/StreletsUpwindEdgeAlg.C:45]: @rcknaus I started looking into this, but it wasn't quite as straight forward as I thought to detect that upwind parameters were being parsed/set after checking the turbulence model.  Probably not hard to implement, but I think the version I have now is adequate for the intended purposes.  

Also, after reviewing the upwinding again I'm not really sure we need to error out over these parameters since @gantech's original implementation included the limiters in the upwind calculation.  The upwinding parameters are pretty much what we usually run anyways. Personally, I'd rather change this to a warning instead of an error.

File: src/edge_kernels/MomentumOpenEdgeKernel.C
Comment (psakievich) [reply src/edge_kernels/MomentumOpenEdgeKernel.C]: @rcknaus @sayerhs This is one thing I haven't looked too much into.  Not sure if it is truly necessary either. @gantech sent me some of his calculations, but I haven't had a chance to review them yet.  If we want to hold up the merge I'll take a closer look.  The regression test doesn't exercise this capability and neither do any of the tests we've run for the Q1 milestone to-date. So far they have all been periodic bc's.

**Comment by psakievich** [comment]
I've addressed just about everything here. There are a couple open items that I've commented on and left unresolved in the discussion above.  Also, I decided to go ahead and pull the SST_IDDES_ABL model from this PR. I felt I wasn't really doing a good job paying attention to that case and I'm less aware of the details.  Also should make this PR smaller. My thought is it can be added back in a separate PR after we've validated the base SST_IDDES model.  

The regression test has a very similar convergence to the `SSTPeriodicHillEdge` test and the results look pretty much the same so I believe we should be good to go. @lawrenceccheung we should still do some extra validation on the Q1 problems since I didn't run this for very long.

**Pull request merged**
